File: backend/app/schemas/bolge.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from .geo import MAKS_HALKA_NOKTASI

# Renk her zaman #rrggbb (frontend'deki renk secici bunu uretir).
RENK_DESENI = r"^#[0-9a-fA-F]{6}$"


def _nokta_sayisi_sinirli(noktalar: list[list[tuple[float, float]]]) -> None:
    """Kaydedilen geometrinin dizi basina nokta ust siniri (bkz. schemas/geo.py).
    Bolgeler kalici olarak yazildigi icin sinir burada ayrica gerekli: sinirsiz
    bir geometri hem PostGIS'i hem sonraki her okumayi surekli mesgul ederdi."""
    for dizi in noktalar:
        if len(dizi) > MAKS_HALKA_NOKTASI:
            raise ValueError(
                f"bir nokta dizisi en fazla {MAKS_HALKA_NOKTASI} nokta icerebilir"
            )


class BolgeGirdi(BaseModel):
    """Yeni bir gorev bolgesi kaydi.

    `noktalar` bir LISTE LISTESIDIR (sinirlar API'siyle ayni sekil): her eleman
    bir poligon halkasi ([lon,lat] cifti listesi). Halkalarin kapali olmasi
    gerekmez, backend kapatir."""

    ad: str = Field(min_length=1, max_length=120)
    aciklama: str | None = Field(default=None, max_length=1000)
    renk: str = Field(default="#059669", pattern=RENK_DESENI)
    # Kaydi sahiplenecek mudurluk. YALNIZCA ADMIN icin anlamlidir: departmani
    # olan personelin kaydi her zaman kendi mudurlugune yazilir, gonderdigi
    # deger yok sayilir (bkz. routers/bolgeler.py). None = genel.
    departman: str | None = Field(default=None, max_length=32)
    noktalar: list[list[tuple[float, float]]]

    @model_validator(mode="after")
    def noktalar_gecerli_olmali(self) -> "BolgeGirdi":
        if not self.noktalar:
            raise ValueError("en az bir nokta dizisi gonderilmelidir")
        for halka in self.noktalar:
            if len(halka) < 3:
                raise ValueError("her alan halkasi en az 3 nokta icermelidir")
        _nokta_sayisi_sinirli(self.noktalar)
        return self


class BolgeGuncelle(BaseModel):
    """Kaydedilmis bir bolgenin guncellenmesi: ad/aciklama/renk ve GEOMETRI.

    `noktalar` verilirse kaydin sekli degisir (haritada koseler surukleyerek /
    kenara nokta ekleyerek / alani genisletip daraltarak duzenlenir). Sekil
    yerinde guncellenir: kaydin kimligi, atamasi ve gecmisi korunur - bir
    bolgenin sinirlarini duzeltmek yeni bir bolge acmayi gerektirmesin."""

    ad: str | None = Field(default=None, min_length=1, max_length=120)
    aciklama: str | None = Field(default=None, max_length=1000)
    renk: str | None = Field(default=None, pattern=RENK_DESENI)
    # Kaydi baska bir mudurluge devretmek (yalnizca admin). `exclude_unset` ile
    # ayirt edilir: `departman: null` "genel yap", alani hic gondermemek
    # "dokunma" demektir - UserUpdate.yaka ile ayni desen.
    departman: str | None = Field(default=None, max_length=32)
    noktalar: list[list[tuple[float, float]]] | None = None

    @model_validator(mode="after")
    def noktalar_bos_olmamali(self) -> "BolgeGuncelle":
        if self.noktalar is not None:
            if not self.noktalar:
                raise ValueError("en az bir nokta dizisi gonderilmelidir")
            _nokta_sayisi_sinirli(self.noktalar)
        return self


class BolgeAtama(BaseModel):
    """Bolgeyi bir saha ekibine atar; worker_id=None atamayi kaldirir."""

    worker_id: uuid.UUID | None = None


class BolgeCikti(BaseModel):
    id: uuid.UUID
    ad: str
    aciklama: str | None
    renk: str
    # Kaydi sahiplenen mudurluk; None = genel (tum personel gorur). Adi
    # frontend departman sozlugunden cozer - burada tekrarlanmaz.
    departman: str | None = None
    noktalar: list[list[tuple[float, float]]]
    # PostGIS ile hesaplanan gercek (jeodezik) olcu. Frontend'deki duzlemsel
    # yaklastirmadan bagimsiz, kaydedilen deger her zaman ayni cikar.
    alan_m2: float | None = None
    worker_id: uuid.UUID | None = None
    worker_ad: str | None = None
    assigned_at: datetime | None = None
    # Isin dustugu yaka (kaydin temsil noktasindan cozulur: ST_PointOnSurface).
    # Elle atamada "karsi yaka" uyarisi varliklardakiyle ayni bilgiye dayansin
    # diye burada doner.
    yaka: str | None = None
    # Saha ekibi isi bitirdiginde dolar; None ise is devam ediyor demektir.
    tamamlandi_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class BolgeTamamlama(BaseModel):
    """Saha ekibinin bolgeyi tamamlandi isaretlemesi; tamamlandi=False ile
    yanlislikla kapatilan is geri alinir."""

    tamamlandi: bool = True

File: frontend/src/types/bolge.ts
/** Kaydedilmis gorev bolgeleri (alan).
 *  Backend: models/bolge.py, routers/bolgeler.py
 *
 *  Guzergahlar (cizgi) artik AYRI bir tablo ve ayri bir uctur (types/guzergah.ts).
 *  Arayuzde ikisi ayni panellerde/haritada yan yana durdugu icin ortak bir
 *  goruntuleme tipi tanimlanir (`KayitliCizim`): `tip` alani kaydin hangi
 *  uctan geldigini soyler ve iki listenin birlestirildigi yerde (App.tsx'teki
 *  useMemo) yazilir - ONBELLEGE HAM YANIT girer, etiketleme yalnizca tuketim
 *  noktasinda yapilir. */

export const BOLGE_TIPLERI = ["alan", "cizgi"] as const;
export type BolgeTipi = (typeof BOLGE_TIPLERI)[number];

export const BOLGE_TIP_ETIKETLERI: Record<BolgeTipi, string> = {
  alan: "Görev Bölgesi",
  cizgi: "Güzergâh",
};

/** Iki kayit turunun paylastigi alanlar (geometri olcusu disinda hepsi). */
interface KayitliCizimTaban {
  id: string;
  ad: string;
  aciklama: string | null;
  renk: string;
  /** Halka listesi: alanlarda poligon halkalari, guzergahlarda tek elemanli
   *  nokta dizisi. Iki uc ayni sekli dondurur ki cizim/duzenleme mantigi
   *  ayrismasin. */
  noktalar: [number, number][][];
  /** Kaydi sahiplenen mudurluk; null = genel (tum personel gorur). Bir
   *  mudurlugun cizdigi calisma alanini digeri gormemeli, yoksa kendi alakasiz
   *  ekibine atayabilir. Adi departman sozlugunden cozulur. */
  departman: string | null;
  worker_id: string | null;
  worker_ad: string | null;
  assigned_at: string | null;
  /** Isin dustugu yaka; kaydin temsil noktasindan (alan -> icindeki bir nokta,
   *  cizgi -> hattin ortasi) backend'de cozulur. Elle atamada "karşı yaka"
   *  uyarisi varliklardakiyle ayni bilgiye dayansin diye tasinir. */
  yaka: string | null;
  /** Saha ekibi isi bitirdiyse dolu; null ise is devam ediyor. */
  tamamlandi_at: string | null;
  created_at: string;
  updated_at: string;
}

/** `GET /api/bolgeler` yanitinin HAM sekli - onbellege bu girer. */
export interface BolgeYanit extends KayitliCizimTaban {
  /** PostGIS'in jeodezik alan olcusu. */
  alan_m2: number | null;
}

/** Panellerin/haritanin okudugu birlesik kayit: iki uctan gelen satirlar
 *  `tip` ile etiketlenip tek listede toplanir. */
export interface Bolge extends KayitliCizimTaban {
  tip: BolgeTipi;
  /** Alanlarda dolu, guzergahlarda null. */
  alan_m2: number | null;
  /** Guzergahlarda dolu, alanlarda null. */
  uzunluk_m: number | null;
}

export interface BolgeGirdi {
  ad: string;
  aciklama?: string | null;
  renk: string;
  /** Yalnizca ADMIN icin anlamli: departmani olan personelin kaydi her zaman
   *  kendi mudurlugune yazilir, gonderilen deger yok sayilir. */
  departman?: string | null;
  noktalar: [number, number][][];
}

export interface BolgeGuncelle {
  ad?: string;
  aciklama?: string | null;
  renk?: string;
  /** Kaydi baska bir mudurluge devretmek (yalnizca admin); null = genel. */
  departman?: string | null;
  /** Sekil (geometri) guncellemesi - haritada koseler suruklenerek, kenara
   *  nokta eklenerek ya da alan genisletilip daraltilarak degistirilir. */
  noktalar?: [number, number][][];
}

/** Haritada sekli duzenlenmekte olan bir bolge/guzergah. `noktalar` kaydin
 *  kendisinden bagimsiz bir taslaktir: kullanici "Kaydet" demeden backend'e
 *  yazilmaz, "Vazgeç" ile atilir. */
export interface SekilDuzenleme {
  id: string;
  ad: string;
  tip: BolgeTipi;
  renk: string;
  noktalar: [number, number][][];
}
